### PR TITLE
Decode sampling on device for benchmark LLMs

### LINKED ISSUE #3617: Keep all tensors on device between LLM decode iterations
## Summary

During LLM decode, every iteration round-trips tensors through the CPU host. This is a significant performance bottleneck, especially for multi-chip setups where device-to-host transfers are more expensive. Currently, transferring the output tensors to host alone adds ~20ms of overhead per decode step. We should keep all intermediate tensors on device between decode steps by wrapping the LLM in a PyTorch model that performs the post-processing (argmax, cache position update) on-device.

## Current Behavior

In [`generate_and_benchmark`](tests/benchmark/benchmarks/llm_benchmark.py#L182-L255), each decode iteration does:

1. **Forward pass** on device → produces `output` (logits)
2. **Transfer logits to CPU**: `logits = read_logits_fn(output).to("cpu")` (line 218) — **~20ms overhead**
3. **Argmax on CPU**: `next_token_ids = logits[:, -1].argmax(dim=-1)` (line 220)
4. **Transfer next_token_ids back to device**: `next_token_ids.unsqueeze(-1).to(device)` (line 237)
5. **Transfer cache_position to CPU and back**: lines 239-241

This means every single decode step involves **two device↔host round-trips** (logits down, input_ids + cache_position back up), which adds significant latency per token. The output tensor transfer alone (step 2) costs ~20ms, which is a substantial portion of the per-token decode time.

## Proposed Solution

Create a wrapper `nn.Module` around the LLM that includes the post-processing logic (argmax, cache_position increment) as part of the compiled graph, so everything stays on device:

```python
class LLMDecodeWrapper(nn.Module):
    def __init__(self, model, read_logits_fn):
        super().__init__()
        self.model = model
        self.read_logits_fn = read_logits_fn

    def forward(self, input_ids, past_key_values, cache_position, use_cache=True):
        output = self.model(
            input_ids=input_ids,
            past_key_values=past_key_values,
            cache_position=cache_position,
            use_cache=use_cache,
        )
        logits = self.read_logits_fn(output)
        next_token_ids = logits[:, -1].argmax(dim=-1).unsqueeze(-1)
        next_cache_position = cache_position[-1:] + 1
        # Return both the token ids and updated cache position
        # so the next iteration can consume them directly on device
        return next_token_ids, next_cache_position, logits
```

### Key considerations

- **Additional outputs**: The wrapper needs to return `next_token_ids` and `next_cache_position` as additional outputs alongside logits. This changes the compiled graph signature.
- **EOS checking**: Currently EOS is checked every iteration on CPU. With on-device tensors, we'd either check less frequently (e.g. every N steps) or compile the EOS check into the graph as well.
- **PCC validation**: Logits still need to come back to CPU for PCC comparison, but only for the first iteration (validation), not every decode step.
- **Compilation**: The wrapper must be compatible with `torch.compile(backend="tt")`. The argmax and cache_position arithmetic are simple ops that should compile cleanly.
- **Multi-chip sharding**: For tensor-parallel setups, the wrapper outputs need proper sharding marks. The `next_token_ids` and `next_cache_position` should be replicated across devices.

### PR DESCRIPTION
**Pull request by vkovacevicTT**
### Ticket
Closes #3617 

### Problem description
Currently in tt-xla benchmarks, sampling (picking next decode token, updating cache pos) is done on CPU.
This introduces an unnecessary device -> host -> device roundtrip that affects end-to-end performance.

Sampling should be done on device - merged with the model.

Merging torch.argmax with the model introduces two problems:
1.  Keeping logits (which we need for PCC measurement) from all decode steps on device causes DRAM OOM error.
2. `torch.argmax` becomes `ttnn.argmax` that defaults to single_core and is EXTREMELY slow for [batch_size, vocab_size] tensors.

Solution for 1. is to split benchmark into Perf benchmark run and Accuracy (PCC) benchmark run.
In Perf run model would only return (and keep on device) next_token_id.
In PCC run model would return logits, but immediately transfer them to CPU, since we don't care for perf in this run.

Solution for 2. implemented in this PR is to avoid torch.argmax (ttnn.argmax) and use ops that have multicore support:
- **max** - find max logit value for every user/batch
- **eq** - find positions of max values for every user/batch
- **mul** - multiply tensor of indexes with max positions
- **max** - get the last position (last index of token that has max logit value)

This is done at torch level, simply by switching from `torch.argmax` to previously mentioned sequence of ops.

*Tested locally and in CI*
Although device time increases a bit, end-to-end perf improves significantly.

Another solution to explore is enabling multicore argmax through tt-mlir (discover the pattern, [convert to RowMajor](https://github.com/tenstorrent/tt-metal/blob/88f5a99d53df17a7f83ad22331fd595a8fed08f4/ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_device_operation.cpp#L153-L157), ...)
UPDATE:
Switch to this solution when https://github.com/tenstorrent/tt-mlir/pull/7705 lands and tt-mlir gets uplifted.

### What's changed

- introduced `LLMDecodeWrapper` that merges `read_logits_fn` and sampling with model
- introduced `_fast_argmax` for faster sampling on device compared to `torch.argmax`
- split perf and pcc run in benchmark

### Checklist
- [x] [n150 LLMs](https://github.com/tenstorrent/tt-xla/actions/runs/23750430329)
- [x] [LLMBOX LLMs](https://github.com/tenstorrent/tt-xla/actions/runs/23750443431)
- [x] [Galaxy LLMs](https://github.com/tenstorrent/tt-xla/actions/runs/23750458503)
- [ ] [Accuracy run](https://github.com/tenstorrent/tt-xla/actions/runs/23750641953)

### CODE REVIEW

File: tests/benchmark/llm_utils/decode_utils.py [hunk 1]
```diff
@@ -2,31 +2,61 @@
...
+            past_key_values=past_key_values,
+            cache_position=cache_position,
+            use_cache=use_cache,
+        )
+        logits = self.read_logits_fn(output)
+        next_token_ids = logits[:, -1].argmax(dim=-1, keepdim=True)
+        next_cache_position = cache_position[-1:] + 1
+        if self.return_logits:
+            return next_token_ids, next_cache_position, logits
+        return next_token_ids, next_cache_position
```
Comment (odjuricicTT) [inline tests/benchmark/llm_utils/decode_utils.py]: Can you open an issue for us to try to unify these two paths as some point? We could experiment with using 2 command queues, where one would read logits while the other is executing the next decode loop in parallel. In this way we would only need to allocate one logits tensor, fixing the OOM problem.

File: tests/benchmark/llm_utils/decode_utils.py [hunk 2]
```diff
@@ -160,44 +182,37 @@ def generate_and_benchmark(
...
-            logits = read_logits_fn(output).to("cpu")
-            output_logits.append(logits)
-
-            # Greedy decoding: argmax on last position (no sampling/temperature/top_p)
-            next_token_ids = logits[:, -1].argmax(dim=-1)
-
-            # Autoregressive path
-            if ground_truth_tokens is None:
-                output_text = [
-                    tokenizer.decode(token_id) for token_id in next_token_ids
```
[the remark this answers is not in the document]
Comment (vkovacevicTT) [reply tests/benchmark/llm_utils/decode_utils.py]: Correct.
But do we want tokenizer.decode ?
It was just used for verbose print and I think it is unnecessary.
Benchmark logs look much cleaner without it.
